src/conn/conn_compact.c

```diff
@@ -684,15 +684,18 @@ __wt_background_compact_signal(WT_SESSION_IMPL *session, const char *config)
     running = conn->background_compact.running;
 
     WT_ERR(__wt_config_getones(session, config, "background", &cval));
-    if (cval.val == running)
-        /*
-         * This is an error as we are already in the same state and reconfiguration is not allowed.
-         */
+    toggle = cval.val;
+
+    /* Strip the unused fields from the configuration to check if the configuration has changed. */
+    WT_ERR(__wt_config_merge(session, cfg, "background=,exclude=", &stripped_config));
+
+    /* The background compact configuration cannot be changed while it's already running. */
+    if (toggle && running && strcmp(stripped_config, conn->background_compact.config) != 0)
```

Comment:
As discussed, please early exit if we are staying in the same state.

Reply:
At the end of the function you are changing, there are those lines:
```
    if (ret == 0)
        __wt_cond_signal(session, conn->background_compact.cond);
```
I don't think we should execute them either if we are in the same state.

Reply:
That's true, I don't think that's problematic though. Should we change it regardless?
```suggestion
    __wt_spin_unlock(session, &conn->background_compact.lock);
    return (ret);
```

Reply:
I would like to keep the locks in one place, i.e after the `err` label.
@marcbutler, what do you think we should do?

Reply:
Moving it outside the `err` label should be good enough to tackle this.

Reply:
I agree minimizing the paths involved in locking makes it much easier to verify.